src/Felix/Instances/Setoid/Raw.agda

```diff
@@ -0,0 +1,106 @@
+{-# OPTIONS --safe --without-K  #-}
+
+module Felix.Instances.Setoid.Raw where
+
+open import Data.Product using (_,_; _,′_; curry′; uncurry′; ∃₂; proj₁; proj₂)
+open import Data.Product.Function.NonDependent.Setoid using (<_,_>ₛ; proj₁ₛ; proj₂ₛ)
+open import Data.Sum using ([_,_]; inj₁; inj₂)
+open import Data.Sum.Function.Setoid using ([_,_]ₛ; inj₁ₛ; inj₂ₛ)
+open import Data.Sum.Relation.Binary.Pointwise using (inj₁; inj₂)
+open import Data.Unit.Polymorphic using (tt)
+open import Function using (Func) renaming (_∘_ to _∘ᵈ_)
+open import Function.Construct.Composition as Comp
+open import Function.Construct.Constant as Const
+open import Function.Construct.Identity as Id
+open import Level using (_⊔_; suc)
+open import Relation.Binary using (Setoid)
+
+open import Function.Construct.Setoid using (setoid)
+open import Felix.Equiv using (Equivalent)
+open import Felix.Raw
+open import Felix.Instances.Setoid.Type
+  using (module setoid-instances; _⟶_; cong; _⟨$⟩_) public
+open setoid-instances public
+
+module setoid-raw-instances where instance
+
+  equivalent : ∀ {c ℓ} → Equivalent (c ⊔ ℓ) {obj = Setoid c ℓ} _⟶_
+  equivalent = record
+    { _≈_ = λ {From} {To} → Setoid._≈_ (setoid From To)
+    ; equiv = λ {From} {To} → Setoid.isEquivalence (setoid From To)
+    }
+
+  category : ∀ {c ℓ} → Category {obj = Setoid c ℓ} _⟶_
+  category = record
+    { id = Id.function _
+    -- flip′ Comp.function doesn't reduce in goals
+    ; _∘_ = λ g f → Comp.function f g
+    }
+
+  cartesian : ∀ {c ℓ} → Cartesian {obj = Setoid c ℓ} _⟶_
+  cartesian = record
+    { ! = Const.function _ ⊤ tt
+    ; _▵_ = <_,_>ₛ
+    ; exl = proj₁ₛ
+    ; exr = proj₂ₛ
+    }
+
+  cocartesian : ∀ {c ℓ} → Cocartesian ⦃ coproducts {c} {ℓ} ⦄ _⟶_
+  cocartesian = record
+    { ¡ = record
+      { to = λ { () }
+      ; cong = λ { {()} }
+      }
+    ; _▿_ = [_,_]ₛ
+    ; inl = inj₁ₛ
+    ; inr = inj₂ₛ
+    }
+
+  distributive :
+    ∀ {c ℓ} →
+    Distributive
+      ⦃ products {c} {c ⊔ ℓ} ⦄ ⦃ coproducts {c} {ℓ} ⦄
+      _⟶_
+      ⦃ category {c} {c ⊔ ℓ} ⦄ ⦃ cartesian {c} {c ⊔ ℓ} ⦄ ⦃ cocartesian {c} {ℓ} ⦄
+  distributive = record
+    { distribˡ⁻¹ = record
+      { to = λ (a , bc) → [ inj₁ ∘ᵈ (a ,_) , inj₂ ∘ᵈ (a ,_) ] bc
+      ; cong = λ where
+        (a₁≈a₂ , inj₁ b₁≈b₂) → inj₁ (a₁≈a₂ , b₁≈b₂)
+        (a₁≈a₂ , inj₂ c₁≈c₂) → inj₂ (a₁≈a₂ , c₁≈c₂)
+      }
+    ; distribʳ⁻¹ = record
+      { to =  λ (bc , a) → [ inj₁ ∘ᵈ (_, a) , inj₂ ∘ᵈ (_, a) ] bc
+      ; cong = λ where
+        (inj₁ b₁≈b₂ , a₁≈a₂) → inj₁ (b₁≈b₂ , a₁≈a₂)
+        (inj₂ c₁≈c₂ , a₁≈a₂) → inj₂ (c₁≈c₂ , a₁≈a₂)
+      }
+    }
```

Comment:
How tidy could these definitions become if rewritten via the categorical vocabulary?

Reply:
oh, I took me way too long to notice, still can't replace `cong` one as I can't guide Agda to pick proper instances, which I myself don't know in full details. I was hoping agda would figure it out. All I see is the pattern for distribl/r

Reply:
Ok, the issue here is that constructors (`inj₁`/`inj₂`) used at `cong` are coming from [`Pointwise`](https://agda.github.io/agda-stdlib/v2.0/Data.Sum.Relation.Binary.Pointwise.html#998) and we don't have any category instance that would use this as coproducts*. I've tried using `map` from that module but I have trouble convincing Agda is complaining about stuff that I don't know how to solve.

* What would that category be? Category of [`REL A C`](https://agda.github.io/agda-stdlib/v2.0/Relation.Binary.Core.html#780)? [This one?](https://agda.github.io/agda-categories/Categories.Category.Instance.Rels.html)

Reply:
> Ok, the issue here is that constructors (`inj₁`/`inj₂`) used at `cong` are coming from [`Pointwise`](https://agda.github.io/agda-stdlib/v2.0/Data.Sum.Relation.Binary.Pointwise.html#998) [...]

I don't know, and I like this question a lot! If we come up with a good answer, this whole Setoid category might come out beautifully.

Reply:
Note that categorical products and coproducts in `Rels` are necessarily (isomorphic to) *sum* types.